include/pmacc/eventSystem/tasks/TaskReceive.hpp

```diff
@@ -60,25 +60,66 @@ namespace pmacc
+                        }
+                        else
+                        {
+                            exchange->getHostBuffer().setCurrentSize(newBufferSize);
+                            Environment<>::get().Factory().createTaskCopyHostToDevice(
+                                exchange->getHostBuffer(),
+                                exchange->getDeviceDoubleBuffer()
+                            );
+                        }
+
+                        Environment<>::get().Factory().createTaskCopyDeviceToDevice(
+                            exchange->getDeviceDoubleBuffer(),
+                            exchange->getDeviceBuffer(),
+                            this
+                        );
 
-                        Environment<>::get().Factory().createTaskCopyHostToDevice(exchange->getHostBuffer(),
-                                                                                     exchange->getDeviceDoubleBuffer());
-                        Environment<>::get().Factory().createTaskCopyDeviceToDevice(exchange->getDeviceDoubleBuffer(),
-                                                                                       exchange->getDeviceBuffer(),
-                                                                                       this);
                     }
                     else
                     {
-
-                        Environment<>::get().Factory().createTaskCopyHostToDevice(exchange->getHostBuffer(),
-                                                                                     exchange->getDeviceBuffer(),
-                                                                                     this);
+                        if(Environment<>::get().isMpiDirectEnabled())
+                        {
+                            exchange->getDeviceBuffer().setCurrentSize(newBufferSize);
+                            /* We can not be notified from setCurrentSize() therefore
+                             * we need to wait that the current event is finished.
+                             */
+                            setSizeEvent =__getTransactionEvent();
+                            state = WaitForSetSize;
+                        }
+                        else
+                        {
+                            exchange->getHostBuffer().setCurrentSize(newBufferSize);
+                            Environment<>::get().Factory().createTaskCopyHostToDevice(
+                                exchange->getHostBuffer(),
+                                exchange->getDeviceBuffer(),
+                                this
+                            );
+                        }
                     }
+
                     __endTransaction();
                     break;
+                case WaitForSetSize:
+                    // this code is only passed if gpu direct is enabled
+                    if(NULL == Environment<>::get().Manager().getITaskIfNotFinished(setSizeEvent.getTaskId()))
```

Comment:
Perhaps `nullptr` instead of `NULL`? Then also the comparison can be written as `... == nullptr`